src/main/kotlin/com/openlattice/indexing/BackgroundExpiredDataDeletionService.kt

```diff
@@ -0,0 +1,250 @@
+        logger.info(
+                "Starting deletion of expired data for entity set {} with id {}",
+                entitySet.name,
+                entitySet.id
+        )
+        var dataTableDeleteCount = 0
+        if (entitySet.expiration != null) {
+            val pair: Pair<Set<UUID>, Int> = deleteExpiredDataFromDataTable(entitySet.id, entitySet.expiration)
+
+            dataTableDeleteCount = pair.second
+            if (dataTableDeleteCount <= 0) {
+                logger.info("Entity set {} has no expired data", entitySet.name)
+                return dataTableDeleteCount
+            }
+            val expiredEntityKeyIds = pair.first
+            var elasticsearchDataDeleted = false
+            var idsTableDeleteCount = 0
+            if (expiredEntityKeyIds.isNotEmpty()) {
+                elasticsearchDataDeleted = deleteExpiredDataFromElasticSearch(entitySet.id, entitySet.entityTypeId, expiredEntityKeyIds)
+                idsTableDeleteCount = deleteExpiredDataFromIdsTable(expiredEntityKeyIds)
+            }
+
+            //compare deletion from data table and ids table and whether data was deleted from elasticsearch
+            check(expiredEntityKeyIds.size == idsTableDeleteCount) { "Number of entities deleted from data and ids table are not the same. UH OH." } //do something better
+            check(elasticsearchDataDeleted) { "Expired data not deleted from elasticsearch. UH OH." } // also do something better
+            logger.info("Completed deleting {} expired elements from entity set {}.",
+                    idsTableDeleteCount,
+                    entitySet.name)
+        } else {
+            logger.info(
+                    "Entity set {} does not have a data expiration policy.",
+                    entitySet.name
+            )
+        }
+        return dataTableDeleteCount
+    }
+
+    private fun deleteExpiredDataFromDataTable(entitySetId: UUID, expiration: DataExpiration): Pair<Set<UUID>, Int> {
+        val connection = hds.connection
+        val comparisonField: String
+        val expirationField: Any
+        val expirationFieldSQLType: Int
+        var deleteCount = 0;
+        when (expiration.expirationFlag) {
+            ExpirationType.DATE_PROPERTY -> {
+                val propertyTypeId: UUID = expiration.startDateProperty.get()
+                val propertyType = propertyTypes[propertyTypeId]
+                comparisonField = "'$propertyTypeId'"
+                if (propertyType!!.datatype == EdmPrimitiveTypeKind.Date) {
+                    expirationField = OffsetDateTime.ofInstant(Instant.now().minusMillis(expiration.timeToExpiration), ZoneId.systemDefault()).toLocalDate()
+                    expirationFieldSQLType = Types.DATE
+                } else {  //only other TypeKind for date property type is OffsetDateTime
+                    expirationField = OffsetDateTime.ofInstant(Instant.now().minusMillis(expiration.timeToExpiration), ZoneId.systemDefault())
```

Comment:
Since this calculation is done multiple times, it would probably be simpler to add 
`val expirationField = OffsetDateTime.ofInstant(Instant.now().minusMillis(expiration.timeToExpiration), ZoneId.systemDefault())` 
up top, and then if you have a date property type use `expirationField.toLocalDate()`

Reply:
or 
```
val currentInstant = Instant.now().minusMillis(expiration.timeToExpiration).toEpochMilli()
```
and then use that value

Reply:
yes indeed! that is now cleaned up